README


This archive contains the distribution MIDI-Simple-Drummer,
version 0.08:

  An algorithmic MIDI drummer

This software is copyright (c) 2014 by Gene Boggs.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.020.



Changes
Revision history for MIDI-Simple-Drummer

0.08      2014-10-12 21:06:54-04:00 America/New_York
    - Fixed the MIDI time_signature finally!

0.0701    2014-08-30 09:25:54-04:00 America/New_York
    - Gave life to the Euclidean module via J Mates.
    - Added the rotate($necklace) method.
    - Added Euclidean tests!
    - Removed Funky cruft until it's actually ready for prime time.
    - Fixed POD

0.0601    2014-03-03 19:16:51EST-0500 America/New_York
    - Creating eg/back-in-black...
    - Fixed accidental phrase duplication in Rock patterns.
    - Modularized Rock "half beat" code.
    - Added Funk placeholder...

0.06      2014-01-13 22:45:55 America/New_York
    - Enabled MIDI drumkit types (e.g. "power", "brushes").
    - Enabled MIDI reverb and chorus effect setting.
    - Made the rudiment code generically useful.
    - Improved the eg/ rock-drummer and metronome.
    - Set backbeat to "kick snare kick" if the time signature is a multiple of 3.

0.05      2013-12-26 08:09:30 America/New_York
    - Improved documentation!
    - Added the rest of the rudiments.

0.04      2013-12-23 20:26:35 America/Los_Angeles
    - Added bitmask representation handling.
    - Improved documentation!
    - Enabled reverb and chorus CC effects.
    - Added the -pan_width attribute.
    - Enabled the accent_note() method.
    - Added all but one roll and diddle phrases.

0.03      2013.01.27
    - Added time signature support and tests.
    - Replaced duration computaton with the Music:Duration module.
    - Added Changes versioning plugin.
    - Added spellchecking plugin.
    - Removed niave swing() method until true wisdom emerges.
    - Made the eg/basics script more user-friendly.
    - Flesh-out panning features in the Rudiments module.
    - Added a couple more rudiments and better tests.

0.02_1    2012.02.23
    - Added MIDI panning (CC#10).

 ^^^ version mix-up vvv

0.0201    2012.02.05
    - Updated copyright year.
    - Added M-S-D-Rudiments module.
    - Added tests for Rudiments.

 ^^^ version mix-up vvv

0.02      2011.11.17
    - Added multi-track capability, testing and POD enhancement.
    - Added explicit in-line comments.

0.0101    2011.01.05
    - Refactored with Dist::Zilla.

0.01      2011.01.04
    - Added rhythm metric attributes called bars() and swing().
    - It didn't mean a thing when it didn't have that swing.
    - Made tests and an example jazz drummer.
    - Updated the other example programs.
    - Updated POD and comments.
    - Removed API.pod.

0.00_23   2010.06.19
    - How about a Makefile.PL to go with the distribution? Eh?

0.00_22   2010.06.15
    - Fixed -kit vs Kit, -pat vs Pat, mixup. :\
    - Fixed pattern name tests.
    - Began to turn the API POD into a Pod::S5 slideshow.

0.00_21   Fri 2010.05.15
    - Cleaned up docs a bit.

0.00_20   Fri 2010.05.14
    - Normalized things and cleaned up a bit.
    - Made eg/* work again. :)

0.00_19   Fri 2010.05.14
    - Refactored to use proper inheritance for distinct styles.
    - Cleaned up much ugliness.

0.00_18   Wed 2010.04.28
    - Factored-out kits and patterns.
    - Perceived the ugliness that I hacked-up... Must fix ASAP.

0.00_17   Sat Nov 14, 2009
    - Added bpm, volume and accent accessors and an accent method.
    - Spruced-up POD.
    - Added a beat() method TODO test.

0.00_16   Tue Nov 10, 2009
    - Fixed an option_strike() beat call that fubar'd the testing.
    - Fixed mistaken method example syntax in POD.
    - Improved eg/drummer again and made an eg/metronome.

0.00_15   Sat Nov 8, 2009
    - Repaired tests.
    - Improved eg/drummer.
    - Streamlined the pattern/key logic.

0.00_14   Sat Oct 31, 2009
    - Reorganized POD sections.
    - Renamed the fill aptterns to be more meaningful.

0.00_13   Thu Oct 29, 2009
    - Collapsed known pattern code.
    - Improved POD.
    - Made eg/basics a more useful example.

0.00_12   Thu Oct 29, 2009
    - Whoops. Recorrectified.

0.00_11   Thu Oct 29, 2009
    - Smoothified.
    - Improveified.
    - Correctificated.
    - Enhancified.

0.00_10   Mon Oct 19, 2009
    - Note to self: Read the 0.00_09 change notes.
    - Fix eg/drummer.

0.00_09   Mon Oct 19, 2009
    - Note to self: Don't release a distro when you are more tired than alive.

0.00_08   Mon Oct 19, 2009
    - Simplified the public and internal interface.

0.00_07   Sat Oct 17, 2009
    - Something good happened, I'm sure...

0.00_06   Sat Oct 17, 2009
    - Fixed eq/drummer that was "backwards."
    - Made the package synopsis identical to eg/drummer.

0.00_05   Sat Oct 17, 2009
    - Renamed methods.
    - Began adding tests.
    - Enhanced POD.

0.00_04   Tue Oct 13, 2009
    - Allowed user defined patterns to be added to the repertoire.
    - Improved code! Imagine that!
    - Improved documentation.
    - Added POD tests.

0.00_03   Sun Oct 11, 2009
    - Wtf? The _02 distro is blank?

0.00_02   Sun Oct 11, 2009
    - Fixed distribution files and documentation.

0.00_01   Sat Oct 10, 2009
    - First version, released on an unsuspecting world by Module::Starter.
    - Made it live in the Possible Now.



Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker ;



my %WriteMakefileArgs = (
  "ABSTRACT" => "An algorithmic MIDI drummer",
  "AUTHOR" => "Gene Boggs <gene\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "MIDI-Simple-Drummer",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "MIDI::Simple::Drummer",
  "PREREQ_PM" => {
    "MIDI::Simple" => 0,
    "Music::Duration" => 0,
    "constant" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.08",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "MIDI::Simple" => 0,
  "Music::Duration" => 0,
  "Test::More" => 0,
  "constant" => 0,
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/MIDI/Simple/Drummer/Funk.pm
package MIDI::Simple::Drummer::Funk;
BEGIN {
  $MIDI::Simple::Drummer::Funk::AUTHORITY = 'cpan:GENE';
}
use strict;
use warnings;
our $VERSION = '0.02';
use parent 'MIDI::Simple::Drummer';

sub new {
    my $self = shift;
    $self->SUPER::new(
        -patch   => 33, # Jazz
        @_
    );
}

sub _default_patterns {
    my $self = shift;
    return {

    };
}

# Custom kit access
sub _default_kit {
    my $self = shift;
    return {
        %{ $self->SUPER::_default_kit() },
        ride1 => ['Ride Cymbal 1'],
        ride2 => ['Ride Cymbal 2'],
        bell  => ['Ride Bell'],
        pedal => ['Pedal Hi-Hat'],
    }
}
sub ride1 { return shift->_set_get('ride1', @_) }
sub ride2 { return shift->_set_get('ride2', @_) }
sub bell  { return shift->_set_get('bell', @_) }
sub pedal { return shift->_set_get('pedal', @_) }

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

MIDI::Simple::Drummer::Funk

=head1 VERSION

version 0.0601

=head1 DESCRIPTION

This package contains a collection of triplet based patterns, loaded by
L<MIDI::Simple::Drummer>.

The constructor can be provided with a specific patch number (default 33 "Jazz
Kit") or the argument C<-brushes =E<gt> 1> to use the "Brushes kit."

Additionally, the methods below are available.

=head1 NAME

MIDI::Simple::Drummer::Funk - Funk drum grooves

=head1 METHODS

=head2 ride1(), ride2(), bell()

Strike (or set) the rides, individually.

=head2 pedal()

"Depress" the pedal hi-hat.

=head1 SEE ALSO

L<MIDI::Simple::Drummer>, the F<eg/*> and F<t/*> test scripts.

=head1 AUTHOR

Gene Boggs <gene@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Gene Boggs.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/MIDI/Simple/Drummer/Jazz.pm
package MIDI::Simple::Drummer::Jazz;
BEGIN {
  $MIDI::Simple::Drummer::Jazz::AUTHORITY = 'cpan:GENE';
}
our $VERSION = '0.02';
use strict;
use warnings;
use parent 'MIDI::Simple::Drummer';

sub new {
    my $self = shift;
    $self->SUPER::new(
        -patch   => 33, # Jazz
        -brushes => 0,
        @_
    );
}

sub _default_patterns {
    my $self = shift;
    return {

1 => sub { # Basic swing with no kick or snare.
    my $self = shift;
    for my $beat (1 .. $self->beats) {
        $self->note($self->TRIPLET_8TH, $self->ride1);
        $self->rest($self->TRIPLET_8TH) for 0 .. 1;
        $self->note($self->TRIPLET_8TH, $self->pedal, $self->ride1);
        $self->rest($self->TRIPLET_8TH);
        $self->note($self->TRIPLET_8TH, $self->ride1);
    }
},

2 => sub { # Syncopated swing with kick and snare.
    my $self = shift;
    for my $beat (1 .. $self->beats) {
        $self->note($self->TRIPLET_8TH, $self->kick, $self->ride1);
        $self->rest($self->TRIPLET_8TH);
        $self->note($self->TRIPLET_8TH, $self->snare);

        $self->note($self->TRIPLET_8TH, $self->snare, $self->pedal, $self->ride1);
        $self->rest($self->TRIPLET_8TH);
        $self->note($self->TRIPLET_8TH, $self->kick, $self->ride1);

        $self->note($self->TRIPLET_8TH, $self->snare, $self->ride1);
        $self->rest($self->TRIPLET_8TH);
        $self->note($self->TRIPLET_8TH, $self->kick);

        $self->note($self->TRIPLET_8TH, $self->kick, $self->pedal, $self->ride1);
        $self->rest($self->TRIPLET_8TH);
        $self->note($self->TRIPLET_8TH, $self->snare, $self->ride1);
    }
},

3 => sub { # Syncopated swing with kick and snare.
    my $self = shift;
    for my $beat (1 .. $self->beats) {
        $self->note($self->TRIPLET_8TH, $self->pedal, $self->ride1);
        $self->rest($self->TRIPLET_8TH);
        $self->note($self->TRIPLET_8TH, $self->snare);

        $self->note($self->TRIPLET_8TH, $self->snare, $self->ride1);
        $self->rest($self->TRIPLET_8TH);
        $self->note($self->TRIPLET_8TH, $self->pedal, $self->ride1);

        $self->note($self->TRIPLET_8TH, $self->snare, $self->ride1);
        $self->rest($self->TRIPLET_8TH);
        $self->note($self->TRIPLET_8TH, $self->pedal);

        $self->note($self->TRIPLET_8TH, $self->kick, $self->pedal, $self->ride1);
        $self->rest($self->TRIPLET_8TH);
        $self->note($self->TRIPLET_8TH, $self->snare, $self->ride1);
    }
},

'1 fill' => sub {
    my $self = shift;
    $self->note($self->TRIPLET_8TH, $self->snare);
    $self->rest($self->TRIPLET_8TH);
    $self->note($self->TRIPLET_8TH, $self->snare);

    $self->note($self->_8TH, $self->snare) for 0 .. 1;
},

'2 fill' => sub {
    my $self = shift;
    $self->note($self->TRIPLET_8TH, $self->snare);
    $self->rest($self->TRIPLET_8TH);
    $self->note($self->TRIPLET_8TH, $self->snare);

    $self->note($self->TRIPLET_8TH, $self->snare);
    $self->rest($self->TRIPLET_8TH);
    $self->note($self->TRIPLET_8TH, $self->snare);

    $self->note($self->_8TH, $self->snare) for 0 .. 1;
},

'3 fill' => sub { # Ala Buddy
    my $self = shift;
    $self->note($self->TRIPLET_8TH, $self->snare);
    $self->note($self->TRIPLET_8TH, $self->kick);
    $self->note($self->TRIPLET_8TH, $self->snare);

    $self->note($self->TRIPLET_8TH, $self->snare);
    $self->note($self->TRIPLET_8TH, $self->kick) for 0 .. 1;


    $self->note($self->TRIPLET_8TH, $self->snare);
    $self->note($self->TRIPLET_8TH, $self->kick) for 0 .. 1;

    $self->note($self->TRIPLET_8TH, $self->snare);
    $self->note($self->TRIPLET_8TH, $self->kick) for 0 .. 1;
},

    };
}

# Custom kit access
sub _default_kit {
    my $self = shift;
    return {
        %{ $self->SUPER::_default_kit() },
        ride1 => ['Ride Cymbal 1'],
        ride2 => ['Ride Cymbal 2'],
        bell  => ['Ride Bell'],
        pedal => ['Pedal Hi-Hat'],
    }
}
sub ride1 { return shift->_set_get('ride1', @_) }
sub ride2 { return shift->_set_get('ride2', @_) }
sub bell  { return shift->_set_get('bell', @_) }
sub pedal { return shift->_set_get('pedal', @_) }

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

MIDI::Simple::Drummer::Jazz

=head1 VERSION

version 0.08

=head1 DESCRIPTION

This package contains a collection of triplet based patterns, loaded by
L<MIDI::Simple::Drummer>.

The constructor can be provided with a specific patch number (default 33 "Jazz
Kit") or the argument C<-brushes =E<gt> 1> to use the "Brushes kit."

Additionally, the methods below are available.

=head1 NAME

MIDI::Simple::Drummer::Jazz - Jazz drum grooves

=head1 METHODS

=head2 ride1(), ride2(), bell()

Strike (or set) the rides, individually.  By default, these are the
kit rides.  Imagine that!

=head2 pedal()

"Depress" the pedal hi-hat.

=head1 SEE ALSO

L<MIDI::Simple::Drummer>, the F<eg/*> and F<t/*> test scripts.

=head1 AUTHOR

Gene Boggs <gene@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Gene Boggs.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/MIDI/Simple/Drummer/Rock.pm
package MIDI::Simple::Drummer::Rock;
BEGIN {
  $MIDI::Simple::Drummer::Rock::AUTHORITY = 'cpan:GENE';
}
use strict;
use warnings;
use parent 'MIDI::Simple::Drummer';
our $VERSION = '0.03';

use constant CLOSED => 'Closed Hi-Hat';
use constant CRASH  => 'Crash Cymbal 1';
use constant BELL   => 'Ride Bell';
use constant RIDE2  => 'Ride Cymbal 2';
use constant TOM1   => 'High Tom';
use constant TOM2   => 'Low-Mid Tom';
use constant TOM3   => 'Low Tom';
use constant TOM4   => 'High Floor Tom';
use constant TOM5   => 'Low Floor Tom';

sub new {
    my $self = shift;
    $self->SUPER::new(
        -patch => 1, # Standard
        -power => 0,
        -room  => 0,
        @_
    );
}

# "Quater-note beat" Qn tick. Cym on 1. Kick 1&3. Snare 2&4.
sub _quarter {
    my $self = shift;
    my %args = @_;
    for my $beat (1 .. $self->beats) {
        $self->note($self->QUARTER,
            $self->rotate($beat, $args{-rotate}),
            $self->strike($args{-key_patch})
        );
    }
}

# Common eighth-note based grooves with kick syncopation.
sub _eighth {
    my $self = shift;
    my %args = @_;
    for my $beat (1 .. $self->beats) {
        # Lay-down a back-beat rhythm.
        $self->note($self->EIGHTH,
            $self->backbeat_rhythm(%args, -beat => $beat)
        );
        # Add another kick or tick if we're at the right beat.
        $self->note($self->EIGHTH,
            ((grep { $beat == $_ } @{$args{-key_beat}})
                ? ($self->kick, $self->tick) : $self->tick)
        ) if $args{-key_beat} && @{$args{-key_beat}};
    }
}

sub _kick_fill {
    my $self = shift;
    my %args = @_;
    $self->note(
        $args{-note},
        $self->kick,
        $args{-fill}
            ? $self->strike($args{-patches} ? @{$args{-patches}} : CRASH)
            : $self->tick
    );
}

sub _half_one {
    my $self = shift;
    my %args = @_;
    $self->_kick_fill(-note => $self->EIGHTH, %args);
    $self->note($self->EIGHTH, $self->tick);
    $self->note($self->EIGHTH, $self->snare, $self->tick);
    $self->note($self->EIGHTH, $self->tick);
}

sub _half_two {
    my $self = shift;
    my %args = @_;
    $self->note($self->EIGHTH, $self->kick, $self->tick);
    $self->note($self->EIGHTH, $self->kick, $self->tick);
    $self->note($self->EIGHTH, $self->snare, $self->tick);
    $self->note($self->EIGHTH, $self->tick);
}

sub _half_three {
    my $self = shift;
    my %args = @_;
    $self->note($self->EIGHTH, $self->tick);
    $self->note($self->EIGHTH, $self->kick, $self->tick);
    $self->note($self->EIGHTH, $self->snare, $self->tick);
    $self->note($self->EIGHTH, $self->kick, $self->tick);
}

sub _half_four {
    my $self = shift;
    my %args = @_;
    $self->note($self->EIGHTH, $self->kick, $self->tick) for 0 .. 1;
    $self->note($self->EIGHTH, $self->snare, $self->tick);
    $self->note($self->EIGHTH, $self->kick, $self->tick);
}

sub _half_four_two {
    my $self = shift;
    my %args = @_;
    $self->note($self->EIGHTH, $self->tick);
    $self->note($self->EIGHTH, $self->kick, $self->tick);
    $self->note($self->EIGHTH, $self->snare, $self->tick);
    $self->note($self->EIGHTH, $self->tick);
}

sub _half_five {
    my $self = shift;
    my %args = @_;
    $self->note($self->EIGHTH, $self->kick, $self->tick);
    $self->note($self->EIGHTH, $self->tick);
    $self->note($self->EIGHTH, $self->snare, $self->tick);
    $self->note($self->EIGHTH, $self->kick, $self->tick);
}

sub _default_patterns {
    my $self = shift;
    return {

1.1 => sub {
    my $self = shift;
    my %args = (-key_patch => CLOSED, @_);
    $self->_quarter(%args);
},
1.2 => sub {
    my $self = shift;
    my %args = (-key_patch => BELL, @_);
    $self->_quarter(%args);
},
1.3 => sub {
    my $self = shift;
    my %args = (-key_patch => RIDE2, @_);
    $self->_quarter(%args);
},

# Exercises from the see also link.
2.1 => sub { # Ex. 1
    my $self = shift;
    my %args = @_;
    $self->_half_one(%args);
    $self->_half_one;
},
2.2 => sub { # Ex. 2
    my $self = shift;
    my %args = @_;
    $self->_half_one(%args);
    $self->_half_two;
},
2.3 => sub { # Ex. 3
    my $self = shift;
    my %args = @_;
    $self->_half_one(%args);
    $self->_half_three;
},
2.4 => sub { # Ex. 4
    my $self = shift;
    my %args = @_;
    $self->_half_four(%args);
    $self->_half_four_two;
},
2.5 => sub { # Ex. 5
    my $self = shift;
    my %args = @_;
    $self->_half_five(%args);
    $self->_half_four;
},
2.6 => sub { # Ex. 6
    my $self = shift;
    my %args = @_;
    $self->_half_five(%args);
    $self->_half_one;
    $self->_half_one;
    $self->_half_four;
    $self->_half_five;
    $self->_half_one;
},
2.7 => sub { # Ex. 7
    my $self = shift;
    my %args = @_;
    $self->_half_four(%args);
    $self->_half_five;
    $self->_half_five;
    $self->_half_four_two;
    $self->_half_four;
    $self->_half_five;
},
2.8 => sub { # Ex. 8
    my $self = shift;
    my %args = @_;
    $self->_half_one(%args);
    $self->_half_four_two;
    $self->_half_four;
    $self->_half_three;
    $self->_half_one;
    $self->_half_four_two;
},

3.1 => sub { # "Syncopated beat 1" en c-hh. qn k1,3,4&. qn s2,4.
    my $self = shift;
    my %args = (-key_beat => [4], @_);
    $self->_eighth(%args);
},
3.2 => sub { # "Syncopated beat 2" en c-hh. qn k1,3,3&,4&. qn s2,4.
    my $self = shift;
    my %args = (-key_beat => [3, 4], @_);
    $self->_eighth(%args);
},

'1 fill' => sub {
    my $self = shift;
    $self->note($self->SIXTEENTH, $self->snare) for 0 .. 3;
    $self->note($self->SIXTEENTH, $self->strike(TOM1)) for 0 .. 3;
    $self->note($self->SIXTEENTH, $self->strike(TOM3)) for 0 .. 3;
    $self->note($self->SIXTEENTH, $self->strike(TOM5)) for 0 .. 3;
},
'2 fill' => sub {
    my $self = shift;
    $self->note($self->SIXTEENTH, $self->snare, $self->strike(TOM5)) for 0 .. 3;
},
'3 fill' => sub {
    my $self = shift;
    $self->note($self->SIXTEENTH, $self->snare) for 0 .. 1;
    $self->note($self->EIGHTH, $self->snare);
    $self->note($self->SIXTEENTH, $self->strike(TOM1)) for 0 .. 1;
    $self->note($self->EIGHTH, $self->strike(TOM1));
},
'4 fill' => sub {
    my $self = shift;
    $self->note($self->SIXTEENTH, $self->snare, $self->strike(TOM5)) for 0 .. 1;
    $self->note($self->SIXTEENTH, $self->snare) for 0 .. 1;
    $self->note($self->SIXTEENTH, $self->strike(TOM1)) for 0 .. 1;
},
'5 fill' => sub {
    my $self = shift;
    $self->note($self->EIGHTH, $self->kick, $self->strike(CRASH));
    $self->note($self->EIGHTH, $self->snare) for 0 .. 2;
    $self->note($self->SIXTEENTH, $self->strike(TOM1)) for 0 .. 3;
    $self->note($self->SIXTEENTH, $self->strike(TOM3)) for 0 .. 3;
},
'6 fill' => sub {
    my $self = shift;
    $self->note($self->SIXTEENTH, $self->snare) for 0 .. 3;
    $self->note($self->EIGHTH, $self->strike(TOM1));
    $self->note($self->SIXTEENTH, $self->strike(TOM1)) for 0 .. 1;
    $self->note($self->SIXTEENTH, $self->snare) for 0 .. 1;
    $self->note($self->EIGHTH, $self->strike(TOM1));
    $self->note($self->SIXTEENTH, $self->snare) for 0 .. 1;
    $self->note($self->EIGHTH, $self->strike(TOM3));
},
'7 fill' => sub {
    my $self = shift;
    $self->note($self->SIXTEENTH, $self->snare);
    $self->note($self->EIGHTH, $self->snare);
    $self->note($self->SIXTEENTH, $self->strike(TOM1));
    $self->note($self->EIGHTH, $self->strike(TOM1));
    $self->note($self->SIXTEENTH, $self->strike(TOM3)) for 0 .. 1;
    $self->note($self->SIXTEENTH, $self->snare) for 0 .. 3;
    $self->note($self->SIXTEENTH, $self->strike(TOM3)) for 0 .. 1;
    $self->note($self->EIGHTH, $self->strike(TOM3));
},
'8 fill' => sub {
    my $self = shift;
    $self->note($self->EIGHTH, $self->snare);
    $self->note($self->EIGHTH, $self->kick) for 0 .. 1;
    $self->note($self->EIGHTH, $self->snare);
    $self->note($self->EIGHTH, $self->kick) for 0 .. 1;
    $self->note($self->EIGHTH, $self->snare);
    $self->note($self->EIGHTH, $self->kick);
},

    };
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

MIDI::Simple::Drummer::Rock

=head1 VERSION

version 0.08

=head1 DESCRIPTION

This package contains a collection of common rock patterns, loaded by
L<MIDI::Simple::Drummer>.

The constructor can be provided with a specific patch number (default 1
"Standard Kit") or the arguments C<-power =E<gt> 1> or C<-room =E<gt> 1> to use
the alternate rock kits.

=head1 NAME

MIDI::Simple::Drummer::Rock - Rock drum grooves

=head1 TO DO

Some cooler fills, Man.

=head1 SEE ALSO

L<MIDI::Simple::Drummer>, the F<eg/*> and F<t/*> scripts.

Progressive Drum Grooves
L<http://www.amazon.com/Progressive-Drum-Grooves-Series/dp/1875726314/>

=head1 AUTHOR

Gene Boggs <gene@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Gene Boggs.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/MIDI/Simple/Drummer/Euclidean.pm
package MIDI::Simple::Drummer::Euclidean;
BEGIN {
  $MIDI::Simple::Drummer::Euclidean::AUTHORITY = 'cpan:GENE';
}
use strict;
use warnings;
use parent 'MIDI::Simple::Drummer';
our $VERSION = '0.01';

sub new {
    my $self = shift;
    $self->SUPER::new(
        -onsets => 4,
        -patch  => 25,
        -tr808  => 0,
        -rhythm => undef,
        -pad    => 'kick',
        @_
    );
}

sub _default_patterns {
    my $self = shift;
    return {

1 => sub {
    my $self = shift;
    my $rhythm = $self->{-rhythm}
        ? $self->{-rhythm}
        : $self->euclid($self->{-onsets}, $self->beats);
    for my $i ( @$rhythm ) {
        if ( $i eq 'x' ) {
            my $pad = $self->{-pad};
            $self->note($self->EIGHTH, $self->$pad );
        }
        else {
            $self->rest($self->EIGHTH);
        }
    }
}

    };
}

sub euclid {
    my $self = shift;
    my ($m, $n) = @_;

    # Onsets per measure
    $m ||= $self->{-onsets};
    # Beats per measure
    $n ||= $self->beats;

    # Line is from x=0, y=1 to x=$BPM, y=$mod+1
    # Then from that, for each $y from # 1..$mod
    # figure out the x value to see where beat would be.

    my $intercept = 1;

    # y = mx + b; b is 1 as we're drawing the intercept through that point,
    # and then (y2-y1)/(x2-x1) reduces to just:
    my $slope = $m / $n;

    my @onsets = ('.') x $n;
    for my $y ( 1 .. $m ) {
        # solve x = (y-b)/m rounding nearest and put the beat there
        $onsets[ sprintf "%.0f", ( $y - $intercept ) / $slope ] = 'x';
    }

    return \@onsets;
};

sub rotate {
    my $self = shift;
    my $phrase = shift;

    my $done = 0;
    while ( $done == 0 ) {
        my $i = shift @$phrase;
        push @$phrase, $i;
        $done++ if $phrase->[0] eq 'x';
    }

    return $phrase;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

MIDI::Simple::Drummer::Euclidean

=head1 VERSION

version 0.08

=head1 DESCRIPTION

Generate Euclidean rhythms and "world beat" grooves.

=head1 NAME

MIDI::Simple::Drummer::Euclidean - Euclidean Rhythms

=head1 METHODS

=head2 euclid()

Return a list of B<x> onsets or B<.> rests given the C<onsets> and C<beats>.

=head2 rotate()

Rotate the given list of onsets and rests to the next (to the right) onset.

=head1 SEE ALSO

L<MIDI::Simple::Drummer>, the F<eg/*> and F<t/*> scripts.

L<http://student.ulb.ac.be/~ptaslaki/publications/phdThesis-Perouz.pdf>

L<http://cgm.cs.mcgill.ca/~godfried/publications/banff.pdf>

L<http://www.ageofthewheel.com/2011/03/euclidean-rhythm-midi-file-resource-in.html>

=head1 AUTHOR

Gene Boggs <gene@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Gene Boggs.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
